### PR TITLE
OCS Functional Test Suite

### PR DESCRIPTION
**Pull request by davidvossel**
## Functional Tests

Our functional test suite uses the [ginkgo](https://onsi.github.io/ginkgo/) testing framework.

**Prerequisites for running Functional Tests**
- ocs must already be installed
- KUBECONFIG env var must be set

**Running functional test**

```make functest```

Below is some sample output of what to expect.

```
Building functional tests
hack/build-functest.sh
GINKO binary found at /home/dvossel/go/bin/ginkgo
Compiling functests...
    compiled functests.test
Running functional test suite
hack/functest.sh
Running Functional Test Suite
Running Suite: Tests Suite
==========================
Random Seed: 1568299067
Will run 1 of 1 specs

•
Ran 1 of 1 Specs in 7.961 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 0 Skipped
PASS
```

**Developing Functional Tests**

All the functional test code lives in the ```functests/``` directory. For an
example of how a functional test is structured, look at the ```functests/pvc_creation_test.go```
file.

The tests themselves should invoke simple to understand steps. Put any complex
logic in the functests/utils.go package so test flows are easy to follow.

**Running a single test**
When developing a test, it's common to just want to run a single functional test
rather than the whole suite. This can be done using ginkgo's "focus" feature.

All you have to do is put a ```F``` in front of the tests declaration to force
only that test to run. So, if you have an iteration like ```It("some test")```
defined, you just need to set that to ```FIt("some test")``` to force the test
suite to only execute that single test.

Make sure to remove the focus from your test before creating the pull request.
Otherwise the test suite will fail in CI.

### CODE REVIEW

**Comment by davidvossel** [comment]
> I think @jarrpa's request to split up utils.go still remains. Otherwise looks good to me.

Yeah, i broke up the file a bit, but looking back at it I think I can do better. I'll make another update

File: functests/test.go [hunk 1]
```diff
@@ -0,0 +1 @@
+package tests
```
[the remark this answers is not in the document]
Comment (davidvossel) [reply functests/test.go]: i named it functests btw instead of func_tests

File: hack/build-registry-bundle.sh [hunk 2]
```diff
@@ -1,5 +1,7 @@
 #!/bin/bash
 
+set -e
```
Comment (jarrpa) [inline hack/build-registry-bundle.sh]: Do we still want this in the final commit?

Comment (davidvossel) [reply hack/build-registry-bundle.sh]: well, yes. We do want it, but I probably didn't put it in a commit that makes sense for the change.

The issue I hit was 'make ocs-registry' would "pass" even if an error occurred during the docker build process. Without 'set -e', the exit code of `make ocs-registry` doesn't indicate anything useful. 

I'll break it into a separate commit at least.

Comment (davidvossel) [reply hack/build-registry-bundle.sh]: fixed. i broke this into a separate commit.

**Comment by davidvossel** [comment]
> I think the functests uncovered a possible issue. It looks like the StorageCluster is reporting that it is available before the osd's have started. 

I *think* I traced the issue back to how the functional tests wait for the StorageCluster startup condition. If not all expected conditions existed in the list, it looked like it was possible for the "BeforeTestSuite" function to exit earlier than it should have.

I added logic to address this. We'll see if it improves the situation. I was able to verify that the StorageCluster's conditions were set correctly to indicate that it was still progressing when the OSDs were not online. So, it seems likely that what I saw was test related.

**Comment by davidvossel** [comment]
The latest CSV appears to be broken. The functional test failure looks real. 

```
2019-09-13 17:01:13.574026 E | op-osd: failed to create provision job for node example-deviceset-0-jhkfl. Job.batch "rook-ceph-osd-prepare-example-deviceset-0-jhkfl" is invalid: [spec.template.spec.affinity.podAntiAffinity.preferredDuringSchedulingIgnoredDuringExecution[0].podAffinityTerm.topologyKey: Required value: can not be empty, spec.template.spec.affinity.podAntiAffinity.preferredDuringSchedulingIgnoredDuringExecution[0].podAffinityTerm.topologyKey: Invalid value: "": name part must be non-empty, spec.template.spec.affinity.podAntiAffinity.preferredDuringSchedulingIgnoredDuringExecution[0].podAffinityTerm.topologyKey: Invalid value: "": name part must consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character (e.g. 'MyName',  or 'my.name',  or '123-abc', regex used for validation is '([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9]')]
2019-09-13 17:01:13.811323 E | op-osd: failed to create provision job for node example-deviceset-1-bqqv6. Job.batch "rook-ceph-osd-prepare-example-deviceset-1-bqqv6" is invalid: [spec.template.spec.affinity.podAntiAffinity.preferredDuringSchedulingIgnoredDuringExecution[0].podAffinityTerm.topologyKey: Required value: can not be empty, spec.template.spec.affinity.podAntiAffinity.preferredDuringSchedulingIgnoredDuringExecution[0].podAffinityTerm.topologyKey: Invalid value: "": name part must be non-empty, spec.template.spec.affinity.podAntiAffinity.preferredDuringSchedulingIgnoredDuringExecution[0].podAffinityTerm.topologyKey: Invalid value: "": name part must consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character (e.g. 'MyName',  or 'my.name',  or '123-abc', regex used for validation is '([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9]')]
2019-09-13 17:01:15.818636 E | op-osd: failed to create provision job for node example-deviceset-2-6kv9t. Job.batch "rook-ceph-osd-prepare-example-deviceset-2-6kv9t" is invalid: [spec.template.spec.affinity.podAntiAffinity.preferredDuringSchedulingIgnoredDuringExecution[0].podAffinityTerm.topologyKey: Required value: can not be empty, spec.template.spec.affinity.podAntiAffinity.preferredDuringSchedulingIgnoredDuringExecution[0].podAffinityTerm.topologyKey: Invalid value: "": name part must be non-empty, spec.template.spec.affinity.podAntiAffinity.preferredDuringSchedulingIgnoredDuringExecution[0].podAffinityTerm.topologyKey: Invalid value: "": name part must consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character (e.g. 'MyName',  or 'my.name',  or '123-abc', regex used for validation is '([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9]')]

```

and 

```
2019-09-13 17:01:59.350313 E | op-cluster: failed to create cluster in namespace openshift-storage. failed to start the osds. 3 failures encountered while running osds in namespace openshift-storage: failed to create provision job for node example-deviceset-0-jhkfl. Job.batch "rook-ceph-osd-prepare-example-deviceset-0-jhkfl" is invalid: [spec.template.spec.affinity.podAntiAffinity.preferredDuringSchedulingIgnoredDuringExecution[0].podAffinityTerm.topologyKey: Required value: can not be empty, spec.template.spec.affinity.podAntiAffinity.preferredDuringSchedulingIgnoredDuringExecution[0].podAffinityTerm.topologyKey: Invalid value: "": name part must be non-empty, spec.template.spec.affinity.podAntiAffinity.preferredDuringSchedulingIgnoredDuringExecution[0].podAffinityTerm.topologyKey: Invalid value: "": name part must consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character (e.g. 'MyName',  or 'my.name',  or '123-abc', regex used for validation is '([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9]')]
failed to create provision job for node example-deviceset-1-bqqv6. Job.batch "rook-ceph-osd-prepare-examp
```

rook-ceph can't start the osd pods. It seems likely this is some regression in the new rook-ceph image we introduced a few hours ago in this pr, https://github.com/openshift/ocs-operator/pull/121

**Comment by jarrpa** [comment]
How did the test pass on that PR then?

**Comment by davidvossel** [comment]
> How did the test pass on that PR then?

I might have misunderstood the question. If you meant how did the e2e test pass on pr #121 it's because there are no functional tests being executed until this PR is merged. The e2e logic as it is without the functional test suite just verifies that the ocs-operator installs and that the operator is online, not that a storage cluster comes online. That's the part this PR adds, along with verifying PVC creation.

**Comment by jarrpa** [comment]
...ah, right. Okay, we'll probably have to wait until Monday to figure this out. Pinging @kshlm for awareness.

**Comment by davidvossel** [comment]
@jarrpa https://github.com/openshift/ocs-operator/pull/122 should fix this. We weren't setting the topologyKey in our antiaffinity rule that we use on the cephcluster. That causes spec to get rejected during validation by OCP4.

**Comment by jarrpa** [comment]
Gah! I don't know how I missed that one. Merged.

/retest

**Comment by kshlm** [comment]
The tests fail waiting for OSD to come up. From the few test logs I saw, the failure happened just as osds were being created. So maybe bump up the timeout for OSDs.

**Comment by davidvossel** [comment]
> The tests fail waiting for OSD to come up. From the few test logs I saw, the failure happened just as osds were being created. So maybe bump up the timeout for OSDs

I'll up the timeout for now so we can get this PR in. 

Independently of this PR I'm going to circle back around and try to gain an understanding for why this is occurring though. It appears like the StorageCluster CR may be reporting that it's in a stable state (available, upgradable, not progressing, and not degrading) at some point before the OSDs actually come online, but the CR is clearly reflecting the correct status of the StorageCluster when the timeout occurs.

**Comment by davidvossel** [comment]
I bumped the setup timeout to 30 minutes, we'll see if this improves the situation.

**Comment by davidvossel** [comment]
I changed the logic to wait an entire hour for the OSDs to come online, and the e2e test is still timing out.

I'm trying to reproduce this locally using a similar cluster to what the e2e tests use. One difference I noticed is that they're only using two AZs, but after talking with @jarrpa it appears this theoretically this should be fine.

**Comment by davidvossel** [comment]
I was able to reproduce what's happening in Openshift CI by limiting my OCP deployment to only 2 availability zones in AWS.

Only 1 or 2 OSDs ever come online. The OSDs that don't come online complain about not being able to attach a PVC to the OSDs pod. Below is an example.

```
Events:
  Type     Reason              Age                 From                                   Message
  ----     ------              ----                ----                                   -------
  Normal   Scheduled           46m                 default-scheduler                      Successfully assigned openshift-storage/rook-ceph-osd-0-574bb9858b-kfltj to ip-10-0-129-254.ec2.internal
  Warning  FailedAttachVolume  46m                 attachdetach-controller                Multi-Attach error for volume &quot;pvc-230b857f-d8af-11e9-b766-0e15af88a5cc&quot; Volume is already used by pod(s) rook-ceph-osd-prepare-example-deviceset-2-khhxc-22t6n
  Warning  FailedMount         88s (x20 over 44m)  kubelet, ip-10-0-129-254.ec2.internal  Unable to mount volumes for pod &quot;rook-ceph-osd-0-574bb9858b-kfltj_openshift-storage(43ee9141-d8af-11e9-bed1-0a89cbfd94e0)&quot;: timeout expired waiting for volumes to attach or mount for pod &quot;openshift-storage&quot;/&quot;rook-ceph-osd-0-574bb9858b-kfltj&quot;. list of unmounted volumes=[example-deviceset-2-khhxc]. list of unattached volumes=[rook-data rook-config-override rook-ceph-log devices example-deviceset-2-khhxc example-deviceset-2-khhxc-bridge rook-binaries run-udev rook-ceph-osd-token-fcbk9]
```

In order to stabilize the functional tests so we can get at least something merged that verifies basic funcitonality, I'm experimenting with reducing the OSD replica count to 1 just for the functional test environment.

**Comment by davidvossel** [comment]
@jarrpa @obnoxxx Functional tests pass again. I had to deploy a work around for issue #131 

The osd issue occurs when we have multiple worker nodes within the same AZ. Openshift CI gives us only 2 AZs and 3 worker nodes, which means we'll always have at least 1 extra node in an AZ. For the time being, in order to work around #131 **I only label a single node from each AZ and limit the functional test to Replica 2.**

Using Replica 2 for the e2e test suite should only be considered a stop gap. I'm tracking the root cause of the issue in #131 and I've also opened up a discussion with Openshift CI about increasing the AZ count to 3. https://github.com/openshift/release/pull/5047

Here's my suggestion for moving forward.

1. Merge this PR so we have an added level of confidence when developing ocs-operator changes.
2. Consider [removing the e2e test](https://help.github.com/en/articles/enabling-required-status-checks) as a mandatory test. As maintainers, this lets you use the test to gain confidence about risky changes while not being held back by a long running (and complex test) for simple changes. Basically, this is an escape hatch to keep the e2e test from blocking progress. 
3. Change the test to utilize replica 3 once either #131 or https://github.com/openshift/release/pull/5047 are finalized.

**Comment by obnoxxx** [comment]
I'm fine with moving forward like this.

**Comment by obnoxxx** [comment]
@jarrpa @davidvossel :
> Overall looks good. Can you squash the non-vendor commits?

Maybe leave the commit that is taking replica down to 2 separate so that we can revert it later?

**Comment by davidvossel** [comment]
> Maybe leave the commit that is taking replica down to 2 separate so that we can revert it later?

@obnoxxx @jarrpa I squashed down the commits to what I thought looked like good logical chunks. I think where a few changes that made sense as standalone commits.

**Pull request merged**